### ISSUE TITLE
Separating out the unthink-stack to it's own repo.

### ISSUE THREAD

**Issue by tommyphilyaw** [epandco/unthink-cli#65]
I would like to revisit separating out the unthink-stack template into it's own repo. Having it in the CLI repo got the tool out quickly but I think as this stack expands it will be hard to maintain the changes.

A few specific examples that I think would push us to doing this and I think also opens up a broader conversation about the `feature` functionality we want to add to the CLI.

1. Serverless is a goal and as I think through the "base" template it is a bit different than the current `unthink-stack` and would include different scripts. 
2. To that end there would be a different "base" template for Azure and AWS because of how each wants their projects to be. 
3. Similarly in the future say we needed an angular App while the server side pieces may stay largely the same the front end structure would differ and the scripts would too. 

That leads me to that broader conversation where it feels like the CLI repo just needs to the core menu of what "kind" of stack you want and after that the features/menus are provided by a plugin for that "stack". 

So thinking each variation of the stack is it's own repo/plugin for glue gun that offers context specific options. In fact I wonder if we can go a step further in how we break up the pieces to create combinations like so:

Potential repos/plugins:

- unthink-stack-server-express
- unthink-stack-server-aws-lambda
- unthink-stack-server-azure-functions
- unthink-stack-client-riot
- unthink-stack-client-angular

I'm including `angular` only to show variation on the client side with it but the point is that the CLI then could let you build out your stack potentially if we can ensure client/server is decoupled so that you can pick and choose like that. 

The benefit here to me is that each can be maintained and versioned separately and not trying to account for this is one monolithic `unthink-stack` directory in the CLI.  Each plugin would likely have features tailored to it so it makes sense I think to push features down into those as well.

Certainly need more discussion around this just trying to kick it off.

**Comment by lxndrdagreat** [comment]
I agree. We definitely need to seriously plan out how this will work. This ties in to the #30 discussion.

**Comment by lxndrdagreat** [comment]
Could it be as easy as having the CLI download the release ZIP straight from GitHub? It _will_ require being able to _unzip_ on the user's machine... need to look into that.

**Comment by matthewphilyaw** [comment]
My oppositions to downloading zip files is having to determine the latest, pull that down and unzip and manage that process I guess end to end. This would have to include detecting a new version and pull that down before running the tool as well - assuming the goal here would be to have the latest each time. Might be easier than it sounds though so worth exploring. 

As an alternative though could we not simply add these as dependencies to the CLI tool? The massive caveat there is of course the published version on npm is locked down to specific versions of those dependencies unless the user upgrades the tool when a release is done. 

However, with so many moving pieces is that a bad thing? Specifically thinking about the upgrades done in V2 and in the model I proposed above it could touch several of those packages meaning that it would need to release together and the whole tool move forward as a unit to V2. At least it feels that way. 

So by pulling zips down and get latest potentially each time the tool runs could make upgrades harder - maybe?

**Comment by lxndrdagreat** [comment]
I think determining the Zip file would work similarly to the npm package - a particular CLI release is locked into a specific tagged release. The tagged releases are available as archives to download.

That said, I don't have a preference - just trying to provide options.

**Comment by matthewphilyaw** [comment]
If we are OK with the CLI being built to specific versions of those packages (and published) then adding them to the `package.json` for the CLI might be the most straight forward approach because when the user installs the CLI all the packages will be installed with it. 

If the packages are written like gluegun plugins not sure if they are picked up automatically but the very least we will know the package will be installed and under `./node_modules/epandco/unthink-stack-server-express/...` as an example taken from my first post above. Then we can implement what ever scheme we need to I suppose if gluegun's model doesn't fit our needs?

**Comment by matthewphilyaw** [comment]
Maybe the next step we need to take is sketch out what the packages would look like and how they would combine and see what we need to accommodate?

If we take the client and server concepts split them into packages maybe it would be a good starting point to reason through how the packages are structured?

**Comment by lxndrdagreat** [comment]
@matthewphilyaw & @lukecjohnson - Since we now have the server and client layers more-or-less separated out internally, I'd really like to make progress on this front. I'm more than happy to give it a go, but I want to know which route to try: is each piece a Gluegun plugin? Or do we just go with downloading a tagged ZIP of each repo?

Biggest potential problem I see happening with the plugin approach is that it might be tricky to develop/run/test those pieces by themselves. Whereas the ZIP route just seems... archaic to me; but then again maybe the simplest solution is the best?

**Comment by lukecjohnson** [comment]
What about an npm package for each? We could have a monorepo including each template/package - `unthink-stack-server`, `unthink-stack-client`, etc. Then the CLI could basically run a child process to execute something like:
```
npm install @epandco/unthink-stack-server --no-save
```
Then the rest would function the same as the CLI currently does, except it'd look for the template at `node_modules/@epandco/unthink-stack-server`. Then after everything is created, we could delete that folder in `node_modules`. Any potential issues with that approach I'm not thinking of?

**Comment by lukecjohnson** [comment]
Also, with the goal of separating out the client/server layers, how are we going to handle the difference in these templates if they are standalone vs together. For example, if you have a client-only project, the folder structure of `./src/client/...` is unnecessary - it should just be `./src/...`. Similarly, npm scripts like `build-client` could just be `build`.

So are we going to have a standalone server template, standalone client template, and then the full stack template? Or is there another way to handle these differences?

**Comment by lxndrdagreat** [comment]
The gluegun plugin would have to be a NPM package. I don't think there's a reason to make it a NPM package without making it a plugin... I don't see a benefit of that over downloading a ZIP file.

I haven't come up with a solid solve for handling the differences between having one layer (e.g., just the client) vs having more than one (e.g., client and server). I'm not sure that there _is_ a great way to do it. But we're at the point where I think it is more important to take a stab at it and get a foot forward on this, because we've been kicking the can down the road lol.

**Comment by lukecjohnson** [comment]
So with the ZIP approach, we'd just use GitHub releases which would give us a standard URL like https://github.com/epandco/unthink-stack-server/releases/download/1.0.0/unthink-stack-server.zip, right? That should be pretty simple.

But yeah, I think just to get the ball rolling, we could have 3 templates: client only, server only, and full stack. It'd be nice in the future to make it more composable where you can just choose any client template (or not) and choose any server template (or not) and they all mesh together. But for now, having the 3 core templates would cover us in the vast majority of use cases and would definitely be an improvement over what it is now

**Comment by matthewphilyaw** [comment]
@lxndrdagreat I think there are some benefits to making it package, but I think first steps would be to ignore the how for now and like you said just take a stab at separating the pieces to get a feel for it.

Why I think packages would be more beneficial irrespective of use glue plugin:

- Keeping track of the versions more easily. We could just leverage package.json to do it's job and when the CLI is installed it maintains the versions of the referenced packages it was installed with. Could certainly create our own manifest - but just extra work.
- No need to deal with zip files. When the CLI is installed or updated the latest versions specified in it's package.json should be pulled down and available in the node_modules directory. The CLI can focus on copying things out and not having to be package manager too. 

My perspective anyway. 

As far as the pieces being put together maybe each major piece has two flavors of templates: `standalone` | `integrated`:

- The CLI can determine based on context which one to apply
-  `Integrated` flavor would have the template setup to copy into `src` under specific name perhaps. 
- `Standalone` would assume that piece is independent, self contained and would carry the caveat of you can't later add another piece. This is probably OK, because likely wouldn't shift this during a project.
- Each major component would carry the commands relevant for that component I think? So that might actually be a reason to go the package route as well (zips could still work I suppose as long as the CLI points to the right place to find them after the zip file was decompressed). 

This notion of `integrated` vs `standalone` could also solve where to put `templates` in that we don't have universal answer but rather if riot project is `integrated` then it's generated to work with server templates, if it's `standalone` then those templates can be pulled into what ever makes sense for a `standalone` riot project and like wise for the builds. 

Would it be worth the three of us setting up time to work through the problem and sorting out the directory structure?
